Added Cooley-Tukey in Common Lisp

0xJonas opened the pull request:
Essentially a port of my Scala PR (#746) which appears to be stuck in review.

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -0,0 +1,103 @@
+      (cooley-tukey-tailrec data (* 2 stride)))))
+
+(defun cooley-tukey-iter (data)
+  "Performs the DFT using the iterative Cooley-Tukey method."
```
Are we doing a DFT or an FFT, wording gets somewhat confusing here.

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -0,0 +1,103 @@
+      (setf (elt data b-index) difference)))))
+
+(defun cooley-tukey-tailrec (data stride)
+  "Actual iterative implementation of the Cooley-Tukey method."
```
I understand what you're trying to say here with 'iterative,' but it's still recursive. Maybe clarify with a comment. Maybe I'm just nitpicking.

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -0,0 +1,103 @@
+        (return (merge-sub-ffts (cooley-tukey-rec evens)
+                                (cooley-tukey-rec odds))))))
+
+(defun reverse-bits (value num-bits)
```
A very clever way of reversing bits, good job!

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -0,0 +1,103 @@
+
+(defun cooley-tukey-rec (data)
+  "Performs the DFT using the recursive Cooley-Tukey method."
+  (if (<= (length data) 1)
```
The general formatting in use for 'if'-statements is:
```
(if (foo 42)
    (bar 12)
    (baz 103))
```
So add a couple of extra spaces.

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -0,0 +1,103 @@
+
+(defun coefficient (time-index freq-index dft-len)
+  "Calculates a single twiddle factor for the Fourier Transform."
+  (exp (- (/ (* #c( 0 1) 2.0 pi time-index freq-index)
```
One space too many
```suggestion
  (exp (- (/ (* #c(0 1) 2.0 pi time-index freq-index)
```

Trashtalk217 reviewed (changes requested):
An excellent submission, glad that you're interested @0xJonas. I've been the single person implementing the algorithms in Common Lisp up until now and I'm happy to see another Lisper in the archive.

One last thing, It would maybe help to put some extra comments throughout the code. The archive tries to provide good code examples for people learning programming languages and it can be helpful to provide some extra information.

The code is still fine as is, however, and if you correct the small formatting problems, it should be ready for merging.

0xJonas commented:
- Changed indentation on the if-statements
- Fixed that one whitespace
- Made iterative implementation use a loop. It is actually much simpler now.
- Added comments

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -14,38 +14,46 @@
+      data
+      (loop
+        for i from 0 below (length data)
+        ;; Split input into even and odd indices into two smaller lists.
```
```suggestion
        ;; Split even and odd indexed elements into two seperate lists.
```

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -14,38 +14,46 @@
-      (logior (ash (reverse-bits lower-half num-low-bits) num-high-bits)
-              (reverse-bits upper-half num-high-bits)))))
+      value
+      ;; Split bits into two parts
```
```suggestion
      ;; Split bits into two parts.
```

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -14,38 +14,46 @@
 (defun merge-sub-ffts (evens odds)
   "Combines the FFTs of the even and odd indices"
```
```suggestion
  "Combines the FFTs of the even and odd indices."
```

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -14,38 +14,46 @@
+         ;; Calculate coefficients for the odd indices.
+         (twiddle-factors (loop for i from 0 below (length odds)
+                             collect (coefficient 1.0 i fft-length)))
+         ;; Multiply values with coefficients
```
```suggestion
         ;; Multiply values with coefficients.
```

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -14,38 +14,46 @@
+                             collect (coefficient 1.0 i fft-length)))
+         ;; Multiply values with coefficients
+         (odd-terms (mapcar #'* odds twiddle-factors)))
+    ;; Combine the two FFTs
```
```suggestion
    ;; Combine the two FFTs.
```

Trashtalk217 commented on contents/cooley_tukey/code/clisp/fft.lisp:
```diff
@@ -14,38 +14,46 @@
+          collect (elt data i) into odds
+        finally
+          ;; Calculate the Fourier Transform for the two smaller lists and
+          ;; combine them into the Fourier Transform for the full input.
```
```suggestion
          ;; combine them into the Fourier Transform of the full input.
```

Trashtalk217 reviewed (changes requested):
Thank you for the quick response, I assumed that you were going to do points for sentences, so I left some suggestions for the comments.

I've rechecked some of the other common lisp code and how comments and docstrings are handled is all over the place (even when one person made all of them), so don't worry about it. Just make sure it's consistent within the file.

0xJonas commented:
Oops, I forgot to update a line number after the previous commit.

Trashtalk217 commented:
Sorry for the late confirmation, I misunderstood something. But it's good to merge!
